Feat!: Support variables in native SQLMesh projects

izeigerman opened the pull request:
This PR introduces support for user-defined variables to SQLMesh native projects. The variables can now be accessed in the following contexts:

* SQL model definitions
* Python model definitions
* Standalone audits
* Native macros
* Jinja macros

Additionally, SQLMesh intelligently detects variable names referenced in all aforementioned contexts and only includes values for referenced variables into snapshots. This means that if a variable changes, only models that actually use it will be impacted.

However, this capability comes with a few constraints, as the usage detection is performed statically by introspecting relevant SQL/Python/Jinja code.

* When using the `@VAR` macro in SQL, the first argument must always be a string literal. The same limitation applies to the `{{ var(...) }}` macro in Jinja code.
* When variables are accessed from within Python models / native macros, either the `context.var` or `evaluator.var` methods must be used, with the `context` / `evaluator` reference names preserved. Using different reference names breaks the variable detection logic, thus preventing the relevant values from being captured into snapshots.

The `@gateway` macro has been refactored to use variables to benefit from the described static analysis.

@treysp please review the docs. Please also note that a dedicated guide will likely be needed for this in future.

tobymao commented on docs/concepts/macros/macro_variables.md:
````diff
@@ -100,3 +100,22 @@ Other macro variables:
+```sql linenums="1"
+SELECT *
+FROM table
+WHERE some_value = @var('<var_name>')
````
why use `@var` and not just the variable name?

izeigerman replied:
More importantly though, having a function allows you to specify a default value in the second argument.

izeigerman replied:
After additional conversation we decided to support both styles

tobymao commented on docs/concepts/models/python_models.md:
```diff
@@ -143,6 +143,28 @@ def execute(
+    start: datetime,
+    end: datetime,
+    execution_time: datetime,
+    **kwargs: t.Any,
```
the purpose of kwargs was to handle vars

izeigerman replied:
It doesn't work though cause it'd impossible to statically detect whether variable is used or not.

izeigerman replied:
I think we can do this if check which variables user specified explicitly in the list of arguments. However, I can see several issues with that:

1. A user might expect variables to be available through `kwargs` by key, but that won't be the case.
2. A user has to make sure to provide a default value for the argument, otherwise it'd fail at runtime when a variable is missing.
3. The interface looks quite different from what we do in macros.

So there's more implicit magic involved imho. Wdyt?

tobymao commented on sqlmesh/core/audit/definition.py:
```diff
@@ -352,7 +354,7 @@ def metadata_hash(self, audits: t.Dict[str, ModelAudit]) -> str:
 
         query = self.render_query(self) or self.query
-        data.append(query.sql(comments=False))
+        data.append(query.sql(dialect=self.dialect, comments=False))
```
this is a breaking change if you do this so need to switch to feat!

maybe should switch this to gen?

tobymao commented on sqlmesh/core/macros.py:
```diff
@@ -866,6 +876,17 @@ def or_(evaluator: MacroEvaluator, *expressions: t.Optional[exp.Expression]) ->
 
+@macro("VAR")
+def var(
+    evaluator: MacroEvaluator, var_name: exp.Expression, default: exp.Expression = exp.null()
```
default should be optional null, python doesn't allow defaults types here because they are mutable and singletons

default = None
default = exp.null() if default is None

izeigerman replied:
I can see us doing exactly this here: https://github.com/TobikoData/sqlmesh/blob/main/sqlmesh/core/macros.py#L822-L823. What's the difference?

tobymao replied:
that's a bug

tobymao commented on sqlmesh/core/model/definition.py:
```diff
@@ -1441,14 +1449,15 @@ def load_sql_based_model(
-        for r in references
-    }
+    jinja_macro_references, used_variables = extract_macro_references_and_variables(
+        *(e.sql(dialect=dialect) for e in pre_statements),
```
gen

treysp commented on docs/concepts/macros/macro_variables.md:
```diff
@@ -100,3 +100,22 @@ Other macro variables:
     * 'evaluating' - The models' logic is being evaluated.
 * @gateway - A string value that represents the name of the selected [gateway](../../guides/connections.md).
+
+## User-defined Variables
```
Since we already have "user-defined macro variables," I think it would help people differentiate if we had another descriptor for these. 

How about "user-defined project variables" since they are a top-level key in the project config?

georgesittas replied:
Maybe "parameters" is a better term here? Wdyt?

treysp replied:
If we're going to use the `@var()` macro name I think it would be good to use "variables" so the connection is obvious. I'm guessing the term "vars" was chosen because that's what dbt uses? 

If we were just going for clarity, I think `@project_param()` or `@project_var()` (with yaml key of same name) would both be  good.

izeigerman replied:
Not a fan of `@project_var`. Too verbose.

georgesittas replied:
Makes sense

github-christophe-oudar commented:
Hello, it looks like a great feature 👍 

I think it could be helpful to clarify few things regarding that feature:
- @DEF is meant to be "model local variable" while @VAR would be "global context variable" right? Then maybe it could be helpful to update and clarify the concepts in https://sqlmesh.readthedocs.io/en/stable/concepts/macros/sqlmesh_macros/#user-defined-variables
- How @VAR are being defined? Through config.yml / config.py? I don't see related documentation in the PR
- Then you would like to set a precedence system with the CLI (to be implemented in a future PR)?

One major pain point in dbt (at least for me) with variables is that any variables change trigger a recompilation of the models as the variables can change the DAG (for instance, `if (var1 == true) FROM my_schema.table1 else FROM my_schema.my_table2`). Then it means that we can't "precompile" the project for production and recompiling it on dbt can take minutes on big projects (because it fetches the schema/tables from the engine). I'm not sure if SQLMesh is affected by the same problem but on runtime variables (such as start/end/runtime dates), you wouldn't change that DAG (right?) and then you could expect the model to be "identical" (and skip any kind of computing).

izeigerman commented:
> How @VAR are being defined? Through config.yml / config.py? I don't see related documentation in the PR

There's a documentation update to the `reference/configuration.md` file.

>  I'm not sure if SQLMesh is affected by the same problem but on runtime variables (such as start/end/runtime dates), you wouldn't change that DAG (right?) and then you could expect the model to be "identical" (and skip any kind of computing).

These variables are not runtime variables. They are set and versioned for each model individually at plan time. Changing a variable will produce new versions for models that use the affected variable.

>  with variables is that any variables change trigger a recompilation of the models as the variables can change the DAG

Yes, that would be the case here as well to an extent. We'd still benefit from caching, but for example, rendered (compiled) queries will need to be rebuilt.

github-christophe-oudar commented:
> There's a documentation update to the reference/configuration.md file.

Alright, I watched the whole configuration page and it's clearer thanks 👍
I'm likely biased but I feel like the template as in https://docs.getdbt.com/reference/dbt_project.yml is easier to understand what's available at which level.

> These variables are not runtime variables. They are set and versioned for each model individually at plan time. Changing a variable will produce new versions for models that use the affected variable.

Thanks that's what I thought it was too 👍 
There are other use cases for actual run time variables but it's likely a "niche" (adhoc templated reports mostly) compared to global variables usage so I think it shouldn't be a priority for you anyway (especially considering you have already start/end/runtime date as runtime ones well integrated).

izeigerman commented:
Thanks, @github-christophe-oudar, appreciate your feedback!

izeigerman commented:
@github-christophe-oudar I think @treysp did a really good job addressing your comments in the other PR: https://github.com/TobikoData/sqlmesh/pull/2396

github-christophe-oudar commented:
Right 👍 
That looks good 🚀

plaflamme commented:
@izeigerman This is great! We've got our own poor-man's version of gateway-scoped variables, but it requires using forward-only plans whenever we make change to the set of variables. So this feature will be welcomed addition!

One thing I'm unclear on is how someone would use this in practice. Is the idea that the same project would have 2 separate configs with different values for these variables?

In our case, we have a set of variables that changes per **gateway**. Each gateway connects to different versions of the same "warehouse": dev, staging and production. The data is different, but everything is expected to be the same. The main reason why we need variables is because the models have to select from external tables which need to be fully qualified and thus are different between each environment, i.e.: `select * from gcp-<dev|stg|prod>-project.dataset.table`.

So ideally for our use-case, `variables` would be defined on the gateways. In fact, when you think about it, the existing `@gateway` variable is such a gateway-scoped variable already: its value is different per gateway, not per project.

Have you considered allowing "Gateway Variables"? So something like `Project Variables < Gateway Variables < Local Variables`.

izeigerman commented:
@plaflamme 

> So ideally for our use-case, variables would be defined on the gateways. In fact, when you think about it, the existing @gateway variable is such a gateway-scoped variable already: its value is different per gateway, not per project.

That's an excellent point! Yes, we're considering gateway variables. So far the only challenge I'm seeing is what should we do in the multi-repo case when different repos might not have gateways configured / named the same way. How do select a correct set of variables for each individual subrepo? Other than that it should be doable.

plaflamme commented:
> what should we do in the multi-repo case when different repos might not have gateways configured / named the same way

That sounds like it would already be a problem, regardless of variables? i.e.: what happens in this situation

```
sqlmesh --gateway=prod -p project/a -p project/b plan
```

and where `project/a` and `project/b` don't agree on a `prod` gateway?

izeigerman commented:
@plaflamme good point! In your example we'll just pick the gateway from the first project specified in the arguments.

z3z1ma commented:
This is sweet 🔥

izeigerman commented:
@plaflamme I've settled on emitting a warning if a gateway with the given name doesn't exist in a project that is part of the multi-repo setup. Otherwise, the gateway variables will be picked up accordingly.

plaflamme commented:
@izeigerman this is great, exactly what we needed, thank you!

tobymao commented on docs/concepts/macros/sqlmesh_macros.md:
````diff
@@ -41,7 +41,74 @@ It uses the following five step approach to accomplish this:
+```sql linenums="1"
+SELECT *
+FROM table
+WHERE int_variable = @INT_VAR
````
do we note anywhere in the docs that vars are always case insensitive? if not i think we should

tobymao commented on sqlmesh/core/macros.py:
```diff
@@ -171,13 +172,14 @@ def evaluate_macros(
                 changed = True
-                if node.name not in self.locals:
+                variables = self.locals.get(c.SQLMESH_VARS, {})
+                if node.name not in self.locals and node.name.lower() not in variables:
```
why is lower() only called on the second one and not the first one?

izeigerman replied:
Seems like we rely on case sensitivity in locals. Some tests failed when I tried that. Investigating this is out of scope of this update.

tobymao replied:
ok

tobymao commented on sqlmesh/core/macros.py:
```diff
@@ -375,6 +377,15 @@ def engine_adapter(self) -> EngineAdapter:
+
+    def var(self, var_name: str, default: t.Optional[t.Any] = None) -> t.Optional[t.Any]:
+        """Returns the value of the specified variable, or the default value if it doesn't exist."""
+        return (self.locals.get(c.SQLMESH_VARS) or {}).get(var_name, default)
```
maybe this method should always normalize input? var_name.lower() so it's consistent

tobymao commented on tests/core/test_model.py:
```diff
@@ -3407,3 +3409,276 @@ def test_end_no_start():
+            kind FULL,
+        );
+
+        SELECT @VAR('test_var_a') AS a, @VAR('test_var_b', 'default_value') AS b, @VAR('test_var_c') AS c, @TEST_MACRO_VAR() AS d;
```
let's add an uppercase test case, @var('TEST_VAR_a')
